implement vertical slider orientation

B-Reif opened the pull request:
This PR implements a vertical orientation for `Slider`. Closes #589.

# What's new
- Sliders demo now includes a radio button toggle for orientation
- New enum `SliderOrientation` with variants `Horizontal` and `Vertical`
- New mutator function `.orientation` to set the slider orientation
- New constructor functions, `new_vertical` and `from_get_set_vertical` to instantiate with vertical orientation.
- Vertical Sliders use alternate keycodes for increment/decrement (up/down instead of right/left).

# What's changed
- Replaces horizontal-specific Slider logic with some functions matching on `orientation` instead. Some functions and parameters are renamed (e.g. `x_range` -> `position_range`).
- Some helper functions are now methods on `Slider` which match on the Slider's orientation.

# Alternatives
- If API breakage is ok, we could add an orientation argument to existing constructor functions (instead of creating new constructor functions).
- We could expose only the `.orientation` mutator function (instead of changing or creating constructor functions).
- We could expose a vertical slider as a different widget. The logic is similar enough that I think the sharing here is appropriate.

emilk commented:
I've skimmed through it, and overall it looks good! Will review properly when I find the time.

emilk commented on egui/src/widgets/slider.rs:
```diff
@@ -85,9 +93,44 @@ impl<'a> Slider<'a> {
     }
 
+    /// Creates a new vertical slider.
+    pub fn new_vertical<Num: emath::Numeric>(
```
It seems unnecessary to have both this and the `.orientation` builder.

Perhaps `Slider::new(value, 0..100).vertical()` is a nicer way to construct a vertical slider anyway? And then we don't need `new_vertical`, `from_get_set_vertical` and `from_get_set_inner`

emilk commented on egui/src/widgets/slider.rs:
```diff
@@ -139,6 +183,11 @@ impl<'a> Slider<'a> {
         self
     }
 
+    pub fn orientation(mut self, orientation: SliderOrientation) -> Self {
```
```suggestion
    /// Vertical or horizontal slider? The default is horizontal.
    pub fn orientation(mut self, orientation: SliderOrientation) -> Self {
```

emilk commented on egui/src/widgets/slider.rs:
```diff
@@ -346,15 +385,85 @@ impl<'a> Slider<'a> {
+                (rect.left() + handle_radius)..=(rect.right() - handle_radius)
+            }
+            SliderOrientation::Vertical => {
+                (rect.top() + handle_radius)..=(rect.bottom() - handle_radius)
```
I think the default for a vertical slider should be higher values higher up, and lower values lower down:

```suggestion
                (rect.bottom() - handle_radius)..=(rect.top() + handle_radius)
```

emilk commented:
<img width="80" alt="Screen Shot 2021-11-07 at 21 05 12" src="https://user-images.githubusercontent.com/1148717/140660189-ec5e21c3-19b5-4752-8ae4-b690aa9776bb.png">

This looks a bit bad to me. I guess most people who use a vertical slider won't want the label text (no shown above), but the slider value would be very nice to show, and should be _below_ the vertical slider.

B-Reif commented:
Ok, made review changes and moved the value below the slider. Also added a new method `add_contents` to invoke inside of either `ui.horizontal` or `ui.vertical` respectively.

![image](https://user-images.githubusercontent.com/12357846/140841666-22d3d597-2161-4bf0-bdc2-97a9e0d29a9a.png)

This implementation will arrange the slider, value, and then label in a vertical layout. We could further group the value and label onto one horizontal line.

B-Reif commented:
I'm having some issues with the keyboard step functionality of Slider, but I experience them on the current master branch also. These could probably be resolved in a separate PR.
